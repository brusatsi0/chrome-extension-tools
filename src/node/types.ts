import type { Node } from 'acorn'
import type { PluginContext, OutputBundle } from 'rollup'
import type { Plugin as VitePlugin, ViteDevServer } from 'vite'
import { ManifestV3 } from './manifest'

export interface AcornLiteral extends Node {
  type: 'Literal'
  raw: string
  value: string
}

export interface CrxPlugin extends VitePlugin {
  /** Runs during dev mode when the file writer has started and server is listening. */
  fileWriterStart?: (
    config: { port: number; outDir: string },
    server: ViteDevServer,
  ) => Promise<void> | void
  /** Runs during the transform hook for the manifest. */
  transformCrxManifest?: (
    this: PluginContext,
    manifest: ManifestV3,
  ) => Promise<ManifestV3 | null | undefined> | ManifestV3 | null | undefined
  /** Runs during generateBundle, before manifest output. */
  renderCrxManifest?: (
    this: PluginContext,
    manifest: ManifestV3,
    bundle: OutputBundle,
  ) => Promise<ManifestV3 | null | undefined> | ManifestV3 | null | undefined
}

// change this to an interface when you want to add options
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type CrxOptions = Record<string, any>

export interface CrxPluginFn {
  (options: CrxOptions): CrxPlugin | CrxPlugin[]
}
