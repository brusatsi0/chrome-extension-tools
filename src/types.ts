import { JsonObject, PackageJson, Promisable } from 'type-fest'
import { isPresent, Unpacked } from './helpers'
import { Plugin } from 'rollup'

type Nullable<TType> = TType | null | undefined
type Manifest = chrome.runtime.Manifest

export type ManifestV2 = Omit<
  chrome.runtime.ManifestV2,
  'name' | 'description' | 'version'
> &
  Partial<
    Pick<
      chrome.runtime.ManifestV2,
      'name' | 'description' | 'version'
    >
  >

export type ManifestV3 = Omit<
  chrome.runtime.ManifestV3,
  'name' | 'description' | 'version'
> &
  Partial<
    Pick<
      chrome.runtime.ManifestV3,
      'name' | 'description' | 'version'
    >
  >

export type ContentScript = Unpacked<
  chrome.runtime.Manifest['content_scripts']
>

export type WebAccessibleResource = Unpacked<
  chrome.runtime.ManifestV3['web_accessible_resources']
>

export type DeclarativeNetRequestResource = {
  id: string
  enabled: boolean
  path: string
}

export function isMV2(
  m?: chrome.runtime.ManifestBase,
): m is chrome.runtime.ManifestV2 {
  if (!isPresent(m)) throw new TypeError('manifest is undefined')
  return m.manifest_version === 2
}

export function isMV3(
  m?: chrome.runtime.ManifestBase,
): m is chrome.runtime.ManifestV3 {
  if (!isPresent(m)) throw new TypeError('manifest is undefined')
  return m.manifest_version === 3
}

export type FileType =
  | 'CSS'
  | 'HTML'
  | 'IMAGE'
  | 'JSON'
  | 'MANIFEST'
  | 'RAW'
  | 'SCRIPT'

export interface BaseAsset {
  fileType: FileType
  /* Input file name, relative to root */
  id: string
  /* Output file name, relative to outDir */
  fileName?: string
  /* Where the file is referenced, html or manifest and JSON path */
  origin?: string
}

export interface StringAsset extends BaseAsset {
  fileType: 'CSS' | 'HTML'
  source?: string
}

export interface RawAsset extends BaseAsset {
  fileType: 'IMAGE' | 'RAW'
  source?: Uint8Array
}

export interface JsonAsset extends BaseAsset {
  fileType: 'JSON'
  source?: JsonObject
}

export interface ManifestAsset extends BaseAsset {
  fileType: 'MANIFEST'
  source?: Manifest
}

export type Asset =
  | StringAsset
  | RawAsset
  | JsonAsset
  | ManifestAsset

interface RPCEHooks {
  manifest?: (
    source: Manifest,
    packageJson: PackageJson,
  ) => Promisable<Nullable<Manifest>>
  html?: (
    source: string,
    file: StringAsset,
  ) => Promisable<Nullable<StringAsset | string>>
  css?: (
    source: string,
    file: StringAsset,
  ) => Promisable<Nullable<StringAsset | string>>
  image?: (
    source: Uint8Array,
    file: RawAsset,
  ) => Promisable<Nullable<RawAsset | Uint8Array>>
  json?: (file: JsonAsset) => Promisable<Nullable<JsonAsset>>
  raw?: (
    source: Uint8Array,
    file: RawAsset,
  ) => Promisable<Nullable<RawAsset | Uint8Array>>
}

export interface RPCEPlugin extends Plugin {
  name: string
  crxTransform?: RPCEHooks
  crxRender?: RPCEHooks
}

export interface ChromeExtensionOptions {
  /**
   * @deprecated This is not supported for MV3, use this instead:
   * ```js
   * import browser from 'webextension-polyfill'
   * ```
   */
  browserPolyfill?:
    | boolean
    | {
        executeScript: boolean
      }
  /**
   * @deprecated Use a dynamic manifest instead.
   * TODO: add link to docs
   */
  extendManifest?:
    | Partial<chrome.runtime.Manifest>
    | (<T extends chrome.runtime.ManifestBase>(manifest: T) => T)
  pkg?: {
    description: string
    name: string
    version: string
  }
}
