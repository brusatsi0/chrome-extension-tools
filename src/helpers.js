import fs from 'fs-extra'
import path from 'path'

export const not = fn => x => !fn(x)

export const loadAssetData = assetPath =>
  fs.readFile(assetPath).then(src => [assetPath, src])

export const zipArrays = (a1, a2) => a1.map((x, i) => [x, a2[i]])

export async function getAssetPathMapFns(assets) {
  return (await assets).map(([assetPath, assetSrc]) => {
    const name = path.basename(assetPath)
    const id = this.emitAsset(name, assetSrc)
    const assetFileName = this.getAssetFileName(id)

    return x => {
      if (typeof x !== 'string') return x

      if (assetPath.endsWith(x)) {
        return assetFileName
      } else {
        return x
      }
    }
  })
}

export const writeFile = dest => ([htmlPath, htmlSrc]) => {
  const baseName = path.basename(htmlPath)
  const destPath = path.join(dest, baseName)
  return fs.writeFile(destPath, htmlSrc)
}
