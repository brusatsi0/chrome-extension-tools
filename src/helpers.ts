import { OutputOptions } from 'rollup'
import { OutputAsset, OutputChunk, OutputBundle } from 'rollup'

export type Unpacked<T> = T extends Array<infer R> ? R : never

export const not =
  <T>(fn: (x: T) => boolean) =>
  (x: T) =>
    !fn(x)

export function isChunk(
  x: OutputChunk | OutputAsset,
): x is OutputChunk {
  return x && x.type === 'chunk'
}

export function isErrorLike(x: unknown): x is Error {
  return typeof x === 'object' && x !== null && 'message' in x
}

export function isOutputOptions(x: any): x is OutputOptions {
  return (
    typeof x === 'object' &&
    !Array.isArray(x) &&
    typeof x.format === 'string' &&
    ['iife', 'es'].includes(x.format)
  )
}

export function isAsset(
  x: OutputChunk | OutputAsset,
): x is OutputAsset {
  return x.type === 'asset'
}

export function isString(x: any): x is string {
  return typeof x === 'string'
}

export function isUndefined(x: unknown): x is undefined {
  return typeof x === 'undefined'
}

export function isNull(x: unknown): x is null {
  return x === null
}

export function isPresent<T>(x: null | undefined | T): x is T {
  return !isUndefined(x) && !isNull(x)
}

export const normalizeFilename = (p: string) =>
  p.replace(/\.[tj]sx?$/, '.js')

export function format(
  strings: TemplateStringsArray,
  ...placeholders: string[]
) {
  const raw = strings.reduce(
    (result, string, i) => result + placeholders[i - 1] + string,
  )
  const formatted = raw.replace(/^  +/gm, '').trim()
  return formatted
}
