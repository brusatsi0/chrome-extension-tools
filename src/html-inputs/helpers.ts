/* ------------- helper functions ------------- */

export const not = <T>(fn: (x: T) => boolean) => (x: T) => !fn(x)
