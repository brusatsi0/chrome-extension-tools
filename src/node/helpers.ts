import { simple } from 'acorn-walk'
import _debug from 'debug'
import fg from 'fast-glob'
import { PluginContext } from 'rollup'
import v8 from 'v8'
import type {
  ManifestV3,
  WebAccessibleResourceById,
  WebAccessibleResourceByMatch,
} from './manifest'
import type { AcornLiteral } from './types'

export const structuredClone = <T>(obj: T): T => {
  return v8.deserialize(v8.serialize(obj))
}

export { _debug } // makes it easy to import w/ intellisense

export const isString = (x: unknown): x is string => typeof x === 'string'

export const isPresent = <T>(x: T): x is NonNullable<T> => !!x

export function isObject<T>(
  value: T,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): value is Extract<T, Record<string, any>> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

export const isResourceByMatch = (
  x: WebAccessibleResourceById | WebAccessibleResourceByMatch,
): x is WebAccessibleResourceByMatch => 'matches' in x

export async function allFiles(manifest: ManifestV3, options: fg.Options) {
  // JSON
  let locales: string[] = []
  if (manifest.default_locale)
    locales = await fg('_locales/**/messages.json', options)

  const rulesets =
    manifest.declarative_net_request?.rule_resources.flatMap(
      ({ path }) => path,
    ) ?? []

  const js = manifest.content_scripts?.flatMap(({ js }) => js) ?? []
  const serviceWorker = manifest.background?.service_worker
  const css = manifest.content_scripts?.flatMap(({ css }) => css)
  const htmlPages = htmlFiles(manifest)

  const icons = [
    Object.values(manifest.icons ?? {}) as string[],
    Object.values(manifest.action?.default_icon ?? {}) as string[],
  ].flat()

  return {
    js: [...new Set(js)].filter(isString),
    css: [...new Set(css)].filter(isString),
    htmlPages: htmlPages,
    icons: [...new Set(icons)].filter(isString),
    locales: [...new Set(locales)].filter(isString),
    rulesets: [...new Set(rulesets)].filter(isString),
    serviceWorker: [serviceWorker].filter(isString),
  }
}

export function htmlFiles(manifest: ManifestV3): string[] {
  const files = [
    manifest.action?.default_popup,
    Object.values(manifest.chrome_settings_overrides ?? {}),
    manifest.devtools_page,
    manifest.options_page,
    manifest.options_ui?.page,
  ]
    .flat()
    .filter(isString)
    .sort()
  return [...new Set(files)]
}

export function decodeManifest(this: PluginContext, code: string): ManifestV3 {
  const tree = this.parse(code)
  let literal: AcornLiteral | undefined
  simple(tree, {
    Literal(n) {
      literal = n as AcornLiteral
    },
  })
  if (!literal) throw new Error('unable to parse manifest code')
  let result: ManifestV3 = JSON.parse(literal.value)
  if (typeof result === 'string') result = JSON.parse(result)
  return result
}

export function encodeManifest(manifest: ManifestV3): string {
  const json = JSON.stringify(JSON.stringify(manifest))
  return `export default ${json}`
}
