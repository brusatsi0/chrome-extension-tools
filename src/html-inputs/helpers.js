import { loadAssetData } from '../helpers'
import { getCssHrefs, getImgSrcs, getJsAssets } from './cheerio'

/* ------------- helper functions ------------- */

export const not = fn => x => !fn(x)

export const isHtml = path => /\.html?$/.test(path)

export const loadHtmlAssets = htmlData =>
  Promise.all(
    htmlData.map(async data =>
      data.concat({
        js: await Promise.all(
          getJsAssets(data).map(loadAssetData),
        ),
        img: await Promise.all(
          getImgSrcs(data).map(loadAssetData),
        ),
        css: await Promise.all(
          getCssHrefs(data).map(loadAssetData),
        ),
      }),
    ),
  )
