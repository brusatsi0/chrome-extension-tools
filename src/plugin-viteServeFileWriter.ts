import fs from 'fs'
import { isUndefined } from 'lodash'
import { Plugin, watch } from 'rollup'
import { from } from 'rxjs'
import { ViteDevServer } from 'vite'
import { interpret } from 'xstate'
import { model as filesModel } from './files.machine'
import { format, isString } from './helpers'
import { join } from './path'
import {
  machine,
  model,
  optionsHook,
  pluginsHook,
  serverHook,
} from './plugin-viteServeFileWriter.machine'
import { excludedHooks } from './plugin-viteServeFileWriter_excludedHooks'
import { isRPCE } from './plugin_helpers'
import { stubId } from './stubId'
import { RPCEPlugin, Writeable } from './types'
import { narrowEvent, useConfig } from './xstate_helpers'

const service = interpret(machine, { devTools: true })
service.subscribe((state) => {
  console.log(
    'vite-serve-file-writer state',
    JSON.stringify(state.value),
  )
})

/**
 * The Vite dev server doesn't write files to the disk,
 * which is a requirement for Chrome Extension development.
 *
 * This plugin uses an instance of Rollup Watch to
 * fulfill this requirement. We'll use a plugin to load
 * files from the dev server, and then write them to disk.
 * This keeps configuration consistent and simple.
 *
 * Since we're still using the Vite server to load files,
 * Vite will call certain hooks (resolveId, load, transform).
 * These hooks should be excluded from Rollup Watch
 * to avoid duplicate hook calls.
 */
export function viteServeFileWriter(): RPCEPlugin {
  let lastError: unknown
  useConfig(service, {
    actions: {
      handleError({ server }, event) {
        const { error } = narrowEvent(event, 'ERROR')
        lastError = error
        server?.close()
      },
    },
    services: {
      rollupWatch:
        ({ options, plugins, server }) =>
        (send) => {
          const watcher = watch({
            ...options,
            // The context should not be touched here
            // we'll rewrite it in the hybrid output plugin
            context: 'this',
            output: {
              ...options?.output,
              dir: server!.config.build.outDir,
            },
            plugins: [
              resolveFromServer(server!),
              // @ts-expect-error Vite is using a different version of Rollup
              ...plugins,
            ],
          })

          watcher.on('event', async (event) => {
            console.log('rollup watcher event', event)
            try {
              if (event.code === 'BUNDLE_END') {
                send(model.events.BUNDLE_END(event))
              } else if (event.code === 'BUNDLE_START') {
                send(model.events.BUNDLE_START(event))
              } else if (event.code === 'ERROR') {
                await event.result?.close()
                const { error } = event

                if (
                  error.message?.includes('is not exported by')
                ) {
                  // TODO: add documentation with example
                  const message = format`Could not complete bundle because Vite did not pre-bundle a dependency.
                    You may need to add this dependency to your Vite config under \`optimizeDeps.include\`.
                    Original Error: ${error.message}`
                  throw new Error(message)
                } else {
                  throw error
                }
              }
            } catch (error) {
              console.log('rollup watcher error')
              delete (error as any).pluginCode
              delete (error as any).frame
              send(model.events.ERROR(error))
            }
          })

          return () => watcher.close()
        },
      waitForServer: ({ server }) =>
        from(
          new Promise<ViteDevServer>((resolve, reject) => {
            if (!server?.httpServer)
              reject(
                new Error(
                  `vite httpServer is ${typeof server?.httpServer}`,
                ),
              )

            server?.httpServer?.once('listening', resolve)
          })
            .then(model.events.SERVER_READY)
            .catch((error) =>
              model.events.ERROR(error, 'waitForServer'),
            ),
        ),
    },
  })

  let isViteServe: boolean
  const pluginName = 'vite-serve-file-writer'
  return {
    name: pluginName,
    config(config, { command }) {
      isViteServe = command === 'serve'
    },
    [pluginsHook](config) {
      if (!isViteServe) return

      /**
       * Vite ignores changes to config.plugins,
       * so we need to modify the plugins in place.
       */
      const externalPlugins = config.plugins as Writeable<
        typeof config.plugins
      >
      /**
       * Rollup Watch will use these plugins to write
       * the files needed for Chrome Extension development
       */
      const internalPlugins: RPCEPlugin[] = []

      externalPlugins.forEach((plugin, i) => {
        if (
          plugin.name.endsWith(pluginName) ||
          plugin.name === 'alias' ||
          plugin.name.startsWith('vite:')
        )
          return

        /**
         * External hooks run when a file is requested from the server
         *
         * Delete hooks that should only be run by internal Rollup Watch
         */
        const externalPlugin = Object.assign(
          {},
          plugin,
        ) as RPCEPlugin
        delete externalPlugin.options
        delete externalPlugin.buildStart
        if (Object.keys(externalPlugin).length > 1)
          externalPlugins[i] = externalPlugin

        const internalPlugin = Object.assign(
          {},
          plugin,
        ) as RPCEPlugin
        // Internal hooks run inside a Rollup Watch instance
        // Delete hooks that should only run on the Vite server
        for (const hookName of excludedHooks)
          delete internalPlugin[hookName]
        if (Object.keys(internalPlugin).length > 1)
          internalPlugins.push(internalPlugin)
      })

      service.start()
      service.send(
        model.events.HOOK_START(pluginsHook, [
          { ...config, plugins: internalPlugins },
        ]),
      )

      // HTML script modules are not emitted in Vite serve
      externalPlugins
        .find(isRPCE)!
        .api.service.send(
          filesModel.events.EXCLUDE_FILE_TYPE('MODULE'),
        )

      if (lastError) throw lastError
    },
    [serverHook](...args: any[]) {
      if (!isViteServe) return

      if (service.initialized) {
        args[0].toJSON = () => 'ViteDevServer'
        service.send(model.events.HOOK_START(serverHook, args))
      }

      if (lastError) throw lastError
    },
    [optionsHook](...args: any[]) {
      if (!isViteServe) return

      if (service.initialized)
        service.send(model.events.HOOK_START(optionsHook, args))
      if (lastError) throw lastError
    },
    closeBundle() {
      service.stop()
    },
  }
}

/**
 * Use Vite's dev server to resolve and load resources.
 *
 * This way we can take advantage of some of Vite's features
 * and speed in background and content scripts.
 */
export function resolveFromServer(
  server: ViteDevServer,
): Plugin {
  return {
    name: 'resolve-from-vite-dev-server',
    resolveId(source) {
      if (source === stubId) return source
      if (source.startsWith('/@fs')) return source

      const id = join(server.config.root, source)
      const fileExists = fs.existsSync(id)
      return fileExists ? id : source
    },
    async load(id) {
      if (id === stubId) return id

      // Add crx query param so plugins can differentiate (eg, exclude from HMR)
      const [baseId, baseParams] = id.split('?')
      const fullParams = new URLSearchParams(baseParams)
      fullParams.set('crx', '')
      const requestId = `${baseId}?${fullParams}`

      const result = await server.transformRequest(requestId)

      if (!result) return null
      if (isString(result)) return result
      if (isUndefined(result.code)) return null

      const { code, map } = result
      return { code, map }
    },
  }
}

/**
 * For use in tests. Resolves when file write operation is complete.
 */
export const filesReady = () =>
  new Promise<void>((resolve, reject) => {
    const sub = service.subscribe({
      next(state) {
        if (state.matches({ serving: 'ready' })) {
          sub.unsubscribe()
          resolve()
        } else if (state.matches('error')) {
          const { error, id } = narrowEvent(state.event, 'ERROR')
          error.id = id
          sub.unsubscribe()
          reject(error)
        }
      },
      error(err) {
        reject(err)
      },
      complete() {
        reject(
          new Error(`The service "${service.id}" has stopped`),
        )
      },
    })
  })
