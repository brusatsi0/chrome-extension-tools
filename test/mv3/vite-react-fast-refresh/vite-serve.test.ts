import { filesReady } from '$src/plugin-viteServeFileWriter'
import { Manifest } from '$src/types'
import { jestSetTimeout } from '$test/helpers/timeout'
import fs from 'fs-extra'
import path from 'path'
import { createServer, ViteDevServer } from 'vite'

jestSetTimeout(30000)

const outDir = path.join(__dirname, 'dist-serve')

let devServer: ViteDevServer
beforeAll(async () => {
  await fs.remove(outDir)

  devServer = await createServer({
    configFile: path.join(__dirname, 'vite.config.ts'),
    envFile: false,
    build: { outDir },
  })
})

afterAll(async () => {
  await devServer.close()
})

test('writes entry points to disk', async () => {
  expect(fs.existsSync(outDir)).toBe(false)

  await Promise.all([devServer.listen(), filesReady()])

  expect(fs.existsSync(outDir)).toBe(true)

  const manifest = 'manifest.json'
  const popup = 'popup.html'

  const manifestPath = path.join(outDir, manifest)
  const manifestSource: Manifest = await fs.readJson(
    manifestPath,
  )
  expect(manifestSource.content_security_policy).toBeUndefined()

  const popupPath = path.join(outDir, popup)
  const popupSource = await fs.readFile(popupPath, 'utf8')
  expect(popupSource).toMatchSnapshot()
})
