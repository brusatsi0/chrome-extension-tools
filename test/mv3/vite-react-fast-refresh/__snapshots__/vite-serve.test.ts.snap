// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writes entry points to disk 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <script type=\\"module\\" src=\\"/@vite/client?delay=popup.html\\"></script>
    <script type=\\"module\\" src=\\"/__inlineScript?inline=popup.html&amp;hash=aCkJlZSfnd&amp;delay=popup.html\\"></script>

    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=1000, initial-scale=1.0\\">
    <title>Popup Page</title>
  </head>
  <body>
    <div id=\\"root\\"></div>
    <script src=\\"/popup.tsx?delay=popup.html\\" type=\\"module\\"></script>
  

</body></html>"
`;
